Improve Trie implementation when different path variables used in same path location

dcr8898 opened the pull request:
@kyleplump and I collaborated on this PR.

## Summary

Router cannot differentiate between multiple defined routes that contain different path variable names in the same location. For example:

```ruby
get "/:id/edit"
get "/:uuid"
```

This PR changes the router behavior to group all variable slugs together for matching purposes and to lazily create and evaluate Mustermann matchers. Discussion below.

The above symptom is illustrated in other open issues, such as #255 (Routes with multiple variables not working in common nested route scenario) and  #256 (Inconsistent trailing slash behaviour). Although trailing slashes are not a documented use case, they sometimes work (and sometimes don't) due to segment-by-segment matching.

The proposed implementation should close #255 , #256 , and #272 , by applying complete route matchers that provide consistent, predictable behavior, as documented in the current Hanami Guides.

We also suggest closing #247 (Failing edge cases for optional variables) as an undocumented use case that is not currently supported. Optional segments are not supported by either the current router or the proposed update.

Closes #247 
Closes #255
Closes #256 
Closes #272 

## Discussion

### Current Router Behavior

Hanami Router 2.0 uses a trie data structure to build a representation of all routes that contain a "dynamic segment," a.k.a., a path variable. It should be noted that all routes that contain only fixed segments are stored in a separate Hash where they can be looked up directly.

Notable features of the current router:

- Routes are stored in the trie as nodes based on each segment of the route (segments defined by the "/" delimiter in the route).
- For path variables/dynamic segments, a Mustermann matcher is created and stored in the trie node, with the matcher used as the key for the node.
- When matching an incoming path to a route, the trie evaluates each segment of the path and calculates and collects derived parameter values *from each dynamic segment as it goes* (this is important).

Because the Mustermann matcher for each dynamic segment is used as the key for that node, different matchers will result in different nodes being created, even if the dynamic segments in question are in parallel locations in different routes.

That is, in the route example above, different matchers would be created for `:id` and `:uuid`, since different path variable names are used.  However, the same node would be have been used for both routes if the same path variable name had been used (say, `:id`). It is worth noting also that different matchers (and therefore different nodes) would also be created if the same path variable names were used, but with different constraints.

The problem arises when matching paths and collecting parameters on the fly, as illustrated below.

The following routes:

```ruby
get "/:id/edit"
get "/:id"
```

Would produce a trie like this:

```mermaid
flowchart LR
  A((root))-->B(((":id")))
  B-->C((("edit")))
```

Where the double circles represent potential path endpoints, so both `:id` and `edit` are potential path endpoints.

In this case, the router has no trouble matching either `/1234` or `/1234/edit` to the correct routes.

However, if the following routes are used:

```ruby
get "/:id/edit"
get "/:uuid"
```

A trie like this will be created:

```mermaid
flowchart LR
  A((root))-->B((":id"))
  B-->C((("edit")))
  A-->D(((":uuid")))
```

Here, `:id` and `:uuid` are different nodes, and `:id` is no longer an endpoint. Now, when the router tries to match a path like "/9d5794e7-12ed-4f00-a6d1-a5be9cd4072e", it has no way to know that this is a UUID and incorrectly matches it to the `:id` node. Crucially, the UUID is collected and stored in the params Hash under the key `:id`. When the router then realizes that this path does not match any route at or below the `:id` node, it has no way to backtrack, "forget" the incorrect `:id` param, and search down a different node path. In fact, backtracking in this way is contrary to trie design and would essentially convert the trie to an ordinary tree data structure.

### Proposed Router Behavior

The proposed router behavior is to store a single variable node for all dynamic segments that are defined in parallel locations in an app's route table. Also, rather than attempting to parse and collect path variable params on the fly, the proposed router delays the creation and evaluation of Mustermann matchers until a potential full path/route match has been identified.

The following routes:

```ruby
get "/:id/edit"
get "/:uuid"
```

Would now create a trie like this:

```mermaid
flowchart LR
  A((root))-->B(((":*")))
  B-->C((("edit")))
```

And each endpoint would contain the information needed to create and evaluate a matcher for a proposed path that evaluates to that endpoint.

## Tests

The proposed implementation adds a new class, Leaf, and refactors the Trie and Node classes. New unit tests are included for each of these classes. New integration tests are also included (in `recognition_spec.rb`). All of these tests are passing.

The proposed implementation breaks two existing tests, [here](https://github.com/hanami/router/blob/e9c15474ec182a1094d933160c6635e7c0afe63c/spec/integration/hanami/router/recognition_spec.rb#L572-L584) and [here](https://github.com/hanami/router/blob/e9c15474ec182a1094d933160c6635e7c0afe63c/spec/integration/hanami/router/scope_spec.rb#L47-L61). These tests each break for reasons that merit further, separate discussion in another comment below, but we believe these tests exercise undocumented behavior and should be changed or deleted. In the alternative, if this behavior is desired, it should be documented and described and we would be happy to try to implement it accordingly.

Interestingly, there is also a commented-out test [here](https://github.com/hanami/router/blob/e9c15474ec182a1094d933160c6635e7c0afe63c/spec/integration/hanami/router/recognition_spec.rb#L552-L570) that documents the current issue, and now passes with the proposed changes. :nerd_face: 

## Conclusion

With guidance, we will delete or modify the breaking tests or, in the alternative, implement whatever documented behavior is desired for these use cases.

Finally, performance is important in this gem. We believe it is possible that the proposed implementation is faster than the current one, but we have not performance tested it. We have no reason to believe it will be slower. Memory usage is also not tested, but we believe fewer objects will be created with the proposed changes.

Thank you!

adam12 commented:
This is a fantastic PR :clap:

I looked at this issue a couple years ago and wasn't able to solve it due to time constraints. I originally tried to keep a stack of each opportunity to branch and then just keep trying each branch from the stack when there was a routing miss. I like your solution a lot better.

Well done! 🥂

dcr8898 commented:
@adam12 Thanks!

We considered using a search algorithm like that, but in addition to the speed issue, you would have to maintain a parallel stack with the params collected on-the-fly (and prune it as needed). That was starting to feel complicated, so we looked for different solutions that would let us keep the speed a trie offers.

Cheers!

dcr8898 commented:
I was looking over the first test discussed above that is broken by the current proposal:

```ruby
    describe "relative variable with permissive constraint" do
      let(:router) do
        described_class.new do
          get ":test", as: :regex, test: /.*/, to: RecognitionTestCase.endpoint("regex")
        end
      end


      it "recognizes route(s)" do
        runner.run!([
          [:regex, "/test/", {test: "test"}]
        ])
      end
    end
```

I noticed that the initial route in this test is also set up in a non-standard (that is, undocumented) way: there is no leading `/`. I think this is a symptom of the practice of using Mustermann matchers in a piecemeal manner on individual path segments: it introduces edge cases and inconsistent behavior by never applying the complete route specification from `routes.rb`.

This problem is illustrated in other open issues, such as #256 (Inconsistent trailing slash behaviour). Although trailing slashes are not a documented use case, they sometimes work (and sometimes don't) due to segment-by-segment matching.

The proposed implementation should close #255, close #273, and close #256, by applying complete route matchers that impose consistent, predictable behavior, as documented in the current Hanami Guides.

I would also suggest closing #247 (Failing edge cases for optional variables) as an undocumented use case that is not currently supported. Optional arguments are not supported by either the current router or the proposed update.

dcr8898 commented:
I see that the CI build is failing at the RSpec step. I don't think this is caused by anything in this pull request.

timriley commented:
Thanks for this amazing PR, folks! And I think it's so cool that the two of you worked together on this 😍 

I want to give this proper attention, so it'll need some brain space. Right now I'm finishing off multiple gateway support for Hanami's DB layer. As soon as I'm done with that, attending to this PR will be my #1 priority. Speak again soon :)

timriley commented:
I haven't forgotten this, folks! I'm releasing rc1 tomorrow and I'll make sure this is reviewed/merged before 2.2.0.

dcr8898 commented:
Let us know if you would like to discuss at all. Questions and explanations might be faster that way. Thanks!

timriley commented:
Update: I've started looking at this!

timriley reviewed (commented):
OK! I'm here! Let's do this.

Firstly, thank you again for your work here, @dcr8898 and @kyleplump! This is a stunningly well put together PR! 🤩

Your explanation and diagrams were extremely helpful for me to understand the issue and the changes you made, and I think your code changes make for a much better overall implementation.

I'd love to see this merged! 💪🏼 

(In fact, I'd also love it if either one of you felt like writing a technical overview document to add to this repo, or peppering these classes with class- and method-level documentation comments explaining how they work. As you no doubt have discovered, it can take a bit of time to understand how everything hangs together, and the more we can do to make that easier for others and our future selves, the better! At this point you two are probably among the foremost experts in this codebase. Let me know what you think!)

Now let me address the matter of the two failing specs:

## First Broken Test

> The first broken test is [here](https://github.com/hanami/router/blob/e9c15474ec182a1094d933160c6635e7c0afe63c/spec/integration/hanami/router/recognition_spec.rb#L572-L584), in `integration/hanami/router/recognition_spec.rb`.

I agree with your recommendations for this. All our public documentation for routes (https://guides.hanamirb.org/v2.1/routing/overview/) shows leading slashes in all cases. That's the behaviour we care about. Let's delete the test and document the change in behaviour.

## Second Broken Test

> The second breaking test is [here](https://github.com/hanami/router/blob/e9c15474ec182a1094d933160c6635e7c0afe63c/spec/integration/hanami/router/scope_spec.rb#L47-L61), in `integration/hanami/router/scope_spec.rb`.

You know what, I think the current passes/failures within this test are what I want! This behaviour feels right to me, and I would be inclined to remove the one failing test case for `GET /it/` (line 59) to allow this test to pass again.

Let me explain my thinking, keeping things mostly at a user perspective.

My first thought relates to the other broken test, above: we do not provide any automatic handling of trailing slashes in Hanami routes. If you want a trailing slash in a route, you have to provide one. Given this, why should we match on the request for `/it/` here? That request has a trailing slash, and we haven't specified one in the routes.

Next, if we look at scopes in general, [our routing guide](https://guides.hanamirb.org/v2.1/routing/overview/) shows them working without the user having to supply any slashes:

```
    scope "about" do
      get "/contact-us", to: "content.contact_us" # => /about/contact-us
      get "/faq", to: "content.faq" # => /about/faq
    end
```

Our tests show a similar story. `spec/integration/hanami/router/scope_spec.rb` has these routes:

```
    scope "trees" do
      root               to: ->(*) { [200, {}, ["Trees (GET)!"]] }
      get     "/cherry", to: ->(*) { [200, {}, ["Cherry (GET)!"]] }
      # ...
    end
```

Leading to these successful requests:

```
  expect(app.request("GET", "/trees", lint: true).body).to eq("Trees (GET)!")
  expect(app.request("GET", "/trees/cherry", lint: true).body).to eq("Cherry (GET)!")
```

From these examples, the way I see it is that using `scope` buys you a URL segment with an implicit leading slash.

From there, our existing documented behaviour is that the root of a scope should match that exact scope with no trailing slash. This is how it works for static scopes now (`scope "trees" do`), and that's what our tests are also showing for a scope with a dynamic segment (`scope ":locale" do`), with the exception of the now-failing `GET /it/` somehow having worked under the prior implementation.

So like the first failing test above, I would chalk this up to undocumented behaviour simply happening to work by coincidence. I think we should drop the one offending line from this test and document the change.

---

The upshot of the decisions here would be that we can very easily move forward with this PR. Delete the first test and delete a single line from this second test and we are good to go!

What do you think?

(I have broader thoughts about how we could make our routes nicer, i.e. doing away with requiring leading slashes altogether, thereby unifying normal routes and scopes, but those are matters for another day!)

If you're happy with the decisions here, I'm happy to make the changes required to this PR and get it merged (though if you want to beat me to it, that's fine too).

I'll also look at the rexml thing separately and sort that out.

Thanks again for your marvellous work!

timriley commented:
@dcr8898 @kyleplump rexml issue is fixed and main branch builds are green and running again.

Part of this fix was https://github.com/hanami/router/commit/d2143c2481027dc1c9f19e59662d6f4bb08dd492, which introduced a merge conflict with this PR. Hopefully a simple one to address though: just dealing with a single "useless assignment" line that I deleted in node.rb.

dcr8898 commented:
Thank you, Tim! I'd like to confer with @kyleplump before replying. We should get back to you shortly!

timriley commented:
Thanks for the update, @dcr8898. One way or another, I _do_ want to get this PR included in 2.2, since it's an important bug fix.

The release is planned for 36 hours from now. I'll hold this PR until the very last minute, if need be, but of course I'd appreciate any feedback from you in the meantime :)

Once again, I'm sorry for the tight timeline here.

dcr8898 commented:
I spoke to @kyleplump before implement the following.

## Changes to tests

- We deleted the failing test from recognition_spec for the reasons discussed.
- I thought I disagreed with you regarding the second failing test in scope_spec, but  I realized I was wrong. :nerd_face:  Therefore, we deleted the single failing expectation in this test.
- Consistent with my separate response above, we removed unneeded tests of duplicate behavior, as you pointed out and recommended.

## Other changes
- Bumped yardoc version references to "2.2.0" for new classes/methods, consistent with your plans for release.
- The conflict resolution is trivial. I will do that in Github shortly.

## Other issues of note

The middleware module in this gem contains a parallel implementation of the routing trie. I did not investigate or make any changes to this module. It probably needs to be refactored in parallel with the changes to the router module.

You mentioned a technical overview doc or comprehensive yardocs. Do you have a preference? If a technical doc, what would that look like? A more comprehensive readme or something else?

Thank you! This was an intimidating project (for me). Kyle's willingness to dive right in helped me tremendously. Once we dug in, it was fun to figure out!

Thank you all! :purple_heart:

timriley commented:
Love it, thank you so much @dcr8898 and @kyleplump! Merging this and getting on with release things tonight. Will follow up about technical docs, etc. at a future time :)
